### PR TITLE
MAINT/BUG: closes overflow bug in stats.tiecorrect

### LINKED ISSUE #5553: stats.tiecorrect overflows
```python
>>> import numpy as np
>>> from scipy import stats
>>> ntie, k = 2000, 2
>>> a = np.repeat(np.arange(k), ntie)
>>> n = a.size  # ntie * k
>>> ranks = stats.rankdata(a)
>>> 1.0 - k * (ntie**3 - ntie) / float(n**3 - n)
0.750000046875003
>>> stats.tiecorrect(ranks)
0.9513266514579157
```

However, if u apply this patch:

```diff
diff --git a/scipy/stats/_rank.pyx b/scipy/stats/_rank.pyx
index 2851baa..a309663 100644
--- a/scipy/stats/_rank.pyx
+++ b/scipy/stats/_rank.pyx
@@ -57,7 +57,7 @@ def tiecorrect(rankvals):

     """
     cdef np.ndarray[np.float64_t, ndim=1] ranks
-    cdef unsigned int i, inext, n, nties, T
+    cdef long long i, inext, n, nties, T
     cdef np.ndarray[np.intp_t, ndim=1] order
     cdef np.ndarray[np.float64_t, ndim=1] sorted_
     cdef double factor
```

result would be correct:

```python
>>> stats.tiecorrect(ranks)
0.750000046875003
```

### PR DESCRIPTION
**Pull request by behzadnouri**
closes https://github.com/scipy/scipy/issues/5553

also, removes the cython code

### CODE REVIEW

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -4591,6 +4589,59 @@ def ks_2samp(data1, data2):
 
     return Ks_2sampResult(d, prob)
 
+
+def tiecorrect(rankvals):
+    """
+    tiecorrect(rankvals)
```
Comment (ev-br) [inline scipy/stats/stats.py]: This is not needed for Python functions

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -4591,6 +4589,59 @@ def ks_2samp(data1, data2):
...
+    >>> tiecorrect(ranks)
+    0.9833333333333333
+
+    """
+    arr = np.sort(rankvals)
+    idx = np.nonzero(np.r_[True, arr[1:] != arr[:-1], True])[0]
+    cnt = np.diff(idx).astype(np.int64)
+
+    if (np.int64(1 << 20) < cnt).any():  # gaurd overflow
+        cnt = cnt.astype(np.float64)
```
Comment (ev-br) [inline scipy/stats/stats.py]: Do you really need two copies here? I mean, a cast to int64 then to float64, both copy.

File: scipy/stats/stats.py [hunk 3]
```diff
@@ -4591,6 +4589,59 @@ def ks_2samp(data1, data2):
...
+    array([ 1. ,  4. ,  2.5,  5.5,  7. ,  8. ,  2.5,  9. ,  5.5])
+    >>> tiecorrect(ranks)
+    0.9833333333333333
+
+    """
+    arr = np.sort(rankvals)
+    idx = np.nonzero(np.r_[True, arr[1:] != arr[:-1], True])[0]
+    cnt = np.diff(idx).astype(np.int64)
+
+    if (np.int64(1 << 20) < cnt).any():  # gaurd overflow
```
Comment (ev-br) [inline scipy/stats/stats.py]: Typo: gaurd

File: scipy/stats/stats.py [hunk 4]
```diff
@@ -4591,6 +4589,59 @@ def ks_2samp(data1, data2):
...
+        cnt = cnt.astype(np.float64)
+
+    size = np.float64(arr.size)
+    return 1.0 if size < 2 else 1.0 - (cnt**3 - cnt).sum() / (size**3 - size)
```
Comment (ev-br) [inline scipy/stats/stats.py]: It's an extreme corner case, of course, but nonetheless: do we have test coverage for size = 0, 1?

**Comment by ev-br** [comment]
A benchmark would be helpful for checking performance.
@WarrenWeckesser do you happen to remember the story behind cython code here?

**Comment by WarrenWeckesser** [comment]
> do you happen to remember the story behind cython code here?

`rankdata` and `tiecorrect` were cythonized to improve their performance and add some fancy ranking options (https://github.com/scipy/scipy/pull/201).  The old code used plain python loops and was very slow.

File: scipy/stats/stats.py (hunk 4 above)
Comment (behzadnouri) [inline scipy/stats/stats.py]: yes [stats/tests/test_rank.py#L13-L23](https://github.com/scipy/scipy/blob/b60f38bda7db007e72937e17f64be10e25977168/scipy/stats/tests/test_rank.py#L13-L23)

**Comment by behzadnouri** [comment]
as for benchmarks, this patch is faster than 0.16.1, and as a bonus you also get correct result:

on `0.16.1`:

```python
In [1]: import numpy as np

In [2]: from scipy.stats import tiecorrect, rankdata

In [3]: np.random.seed(2718281)

In [4]: n = 1 << 20

In [5]: a = np.random.choice(np.random.randn(n // 1000), n)

In [6]: r = rankdata(a)

In [7]: %timeit tiecorrect(r)
10 loops, best of 3: 177 ms per loop

In [8]: tiecorrect(r)   # <<< this number is not accurate!
Out[8]: 0.9999999994352908
```

on branch:
```python
In [9]: %timeit tiecorrect(r)
10 loops, best of 3: 91.9 ms per loop

In [10]: tiecorrect(r)
Out[10]: 0.99999908673916771
```

File: scipy/stats/stats.py [hunk 5]
```diff
@@ -4591,6 +4589,57 @@ def ks_2samp(data1, data2):
...
+    >>> tiecorrect(ranks)
+    0.9833333333333333
+
+    """
+    arr = np.sort(rankvals)
+    idx = np.nonzero(np.r_[True, arr[1:] != arr[:-1], True])[0]
+    cnt = np.diff(idx).astype(np.int64)
+
+    if (np.int64(1 << 20) < cnt).any():  # guard overflow
+        cnt = cnt.astype(np.float64)
```
Comment (ev-br) [inline scipy/stats/stats.py]: My question about memory consumption still stands: this seems to copy twice. Is this really needed, and is there a simple way of avoiding an extra copy? It seems to me one get rid of the cast to `int64` and use something like `(cnt**3).sum(dtype=float) - cnt.sum(dtype=float)`.

**Comment by ev-br** [comment]
Looks good to me modulo a question about what looks like excessive copying.

It would be also a nice bonus to have a look at older tickets (eg https://github.com/scipy/scipy/issues/3611) to see if this patch fixes them as well.

File: scipy/stats/stats.py (hunk 5 above)
Comment (behzadnouri) [inline scipy/stats/stats.py]: @ev-br  `cnt**3` may overflow by itself if some `cnt[i]` is large enough

Comment (larsmans) [inline scipy/stats/stats.py]: `cnt ** 3.` (float exponent) seems to work just fine:

```py
>>> (np.repeat(2**63-1, 1000) ** 3.).sum()
7.846377169233351e+59
>>> float(((2**63-1) ** 3) * 1000)
7.846377169233351e+59
```

Comment (behzadnouri) [inline scipy/stats/stats.py]: @larsmans with float exponent, numpy does the copying and type cast under the hood

Comment (ev-br) [inline scipy/stats/stats.py]: Can we just cast `cnt` to float unconditionally?

Comment (ev-br) [inline scipy/stats/stats.py]: W.r.t. the numpy behavior and copying, I defer to @larsmans

Comment (larsmans) [inline scipy/stats/stats.py]: I'm not perfectly sure about NumPy's implementation of casting in this case, so let's test:

```py
a = np.arange(1<<25)  # 276488k maxresident
a ** 3                # 538940k maxresident
a ** 3.               # 538980k maxresident
a.astype(float) ** 3  # 800948k maxresident
```

Memory figures come from Linux's `/usr/bin/time` on *distinct* test scripts using NumPy 1.8.2. `timeit` shows no difference in speed.

**Comment by ev-br** [comment]
@larsmans are are OK with the last installment,  dd28ee6?

**Comment by larsmans** [comment]
Ok, merged! Thank you all.

**Pull request merged**